pkg/controllers/binding/common.go

```diff
@@ -37,112 +36,98 @@ import (
+	var bindingSpec workv1alpha2.ResourceBindingSpec
+	if len(binding.GetNamespace()) > 0 {
 		bindingObj := binding.(*workv1alpha2.ResourceBinding)
-		targetClusters = bindingObj.Spec.Clusters
-		requiredByBindingSnapshot = bindingObj.Spec.RequiredBy
-		placement = bindingObj.Spec.Placement
-		replicas = bindingObj.Spec.Replicas
-		conflictResolutionInBinding = bindingObj.Spec.ConflictResolution
-		suspension = bindingObj.Spec.Suspension
-	case apiextensionsv1.ClusterScoped:
+		bindingSpec = bindingObj.Spec
+	} else {
 		bindingObj := binding.(*workv1alpha2.ClusterResourceBinding)
-		targetClusters = bindingObj.Spec.Clusters
-		requiredByBindingSnapshot = bindingObj.Spec.RequiredBy
-		placement = bindingObj.Spec.Placement
-		replicas = bindingObj.Spec.Replicas
-		conflictResolutionInBinding = bindingObj.Spec.ConflictResolution
-		suspension = bindingObj.Spec.Suspension
+		bindingSpec = bindingObj.Spec
 	}
 
-	targetClusters = mergeTargetClusters(targetClusters, requiredByBindingSnapshot)
+	targetClusters := calcTargetClusters(bindingSpec.Clusters, bindingSpec.RequiredBy)
 
-	var jobCompletions []workv1alpha2.TargetCluster
-	var err error
-	if workload.GetKind() == util.JobKind {
-		jobCompletions, err = divideReplicasByJobCompletions(workload, targetClusters)
+	jobCompletions, err := divideReplicasByJobCompletions(workload, targetClusters)
+	if err != nil {
+		return err
+	}
+	config := extraConfig{jobCompletions: jobCompletions}
+
+	for _, targetCluster := range targetClusters {
+		err = s.ensureWork(ctx, workload.DeepCopy(), binding, targetCluster, config)
 		if err != nil {
+			klog.Errorf("Failed to ensure work for workload(%s/%s/%s) in cluster %s, err is: %v",
+				workload.GetKind(), workload.GetNamespace(), workload.GetName(), targetCluster.Name, err)
 			return err
 		}
 	}
+	return nil
+}
 
-	for i := range targetClusters {
-		targetCluster := targetClusters[i]
-		clonedWorkload := workload.DeepCopy()
-
-		workNamespace := names.GenerateExecutionSpaceName(targetCluster.Name)
-
-		// If and only if the resource template has replicas, and the replica scheduling policy is divided,
-		// we need to revise replicas.
-		if needReviseReplicas(replicas, placement) {
-			if resourceInterpreter.HookEnabled(clonedWorkload.GroupVersionKind(), configv1alpha1.InterpreterOperationReviseReplica) {
-				clonedWorkload, err = resourceInterpreter.ReviseReplica(clonedWorkload, int64(targetCluster.Replicas))
-				if err != nil {
-					klog.Errorf("Failed to revise replica for %s/%s/%s in cluster %s, err is: %v",
-						workload.GetKind(), workload.GetNamespace(), workload.GetName(), targetCluster.Name, err)
-					return err
-				}
-			}
+// ensureWork ensure Work to be created or updated in the target cluster.
+func (s *workSyncer) ensureWork(ctx context.Context, workload *unstructured.Unstructured, binding metav1.Object, targetCluster workv1alpha2.TargetCluster, config extraConfig) error {
```

Comment:
The `ensureWork` is dedicated to syncing a single work, which is split from the original `ensureWork`.

Reply:
Yes, more dependencies on single work generation will be extended in extraConfig in the future.